### ISSUE TITLE
[Flake] Bigtable table admin snippets crashes

### ISSUE THREAD

**Issue by dbolduc** [googleapis/google-cloud-cpp#7740]
GCB: https://console.cloud.google.com/cloud-build/builds;region=us-east1/db23ecec-bcf4-4063-875f-d70c52b51b15;tab=detail?project=cloud-cpp-testing-resources
Raw: https://storage.googleapis.com/cloud-cpp-community-publiclogs/logs/google-cloud-cpp/main/d85214a59aae81f8330c71964ae569b274372224/fedora-msan-msan/log-db23ecec-bcf4-4063-875f-d70c52b51b15.txt

```
Running GenerateConsistencyToken() example
libc++abi: terminating
*** SIGABRT received at time=1639503165 on cpu 4 ***
PC: @     0x7fa5f8fc92a2  (unknown)  raise
    @     0x7fa5fde468d0        112  absl::lts_20211102::WriteFailureInfo()
    @     0x7fa5fde45ef7        112  absl::lts_20211102::AbslFailureSignalHandler()
    @           0x481727        384  SignalAction()
generated token is : TokenFor-projects/cloud-cpp-testing-resources/instances/test-instance/tables/tbl-2021-12-14-bneu2z8oylx60lsn08rboy9h9gjbpwe95x

Running CheckConsistency() example
Table is consistent with token TokenFor-projects/cloud-cpp-testing-resources/instances/test-instance/tables/tbl-2021-12-14-bneu2z8oylx60lsn08rboy9h9gjbpwe95x

Running DropRowsByPrefix() example
All rows starting with test-prefix/foo/bar/ successfully deleted

Running DropAllRows() example
    @         0xfffffffa  (unknown)  (unknown)
```

The culprit could be any combination of the following:
* #7733, which reimplemented the test in terms of the generated admin APIs
* #7737, which introduced a new bigtable emulator
* #7738, which removed the retry loop around bigtable integration tests. (Potentially, this retry loop has always been covering up these errors)

**Comment by dbolduc** [comment]
One can reproduce this bug more consistently (~90% of the time) by:
1. Making the sample exit faster, by [dropping the processing](https://github.com/googleapis/google-cloud-cpp/blob/d2a68192ebc9c4d3c015e30ca32ad109f32ca9b6/google/cloud/bigtable/examples/table_admin_snippets.cc#L679-L683) after the client call.
2. [Adding a sleep](https://github.com/dbolduc/google-cloud-cpp/blob/d2a68192ebc9c4d3c015e30ca32ad109f32ca9b6/google/cloud/bigtable/examples/table_admin_snippets.cc#L854) after the sample call.

Commenting out the crash handler, we only learn that:
```
Running WaitForConsistencyCheck() example
terminate called without an active exception
```
which implies to me that its an issue with threads not being joined/detached in time. So I am suspicious of `TableAdmin::background_threads_`.

**Comment by devbww** [comment]
> implies to me that its an issue with threads not being joined/detached in time. So I am suspicious of `TableAdmin::background_threads_`.

Yes, I believe you are on the right track.  In particular, because `TableAdmin` owns the `AutomaticallyCreatedBackgroundThreads` it is important that it (well, at least the last `TableAdmin` instance from a set of copies) is not destroyed from one of those background threads, otherwise the `thread::join()` called from `~AutomaticallyCreatedBackgroundThreads()` will throw `resource_deadlock_would_occur`, and then during the unwind we will attempt to destroy an unjoined thread, and that calls `std::terminate()`.  Kaboom!

So, it is important that the `TableAdmin` not be moved into lambda, as that runs the risks having it destroyed from a callback.  I'm guessing that even without the `move` we may have been taking the risk anyway if the compiler saw that `old_admin` was subsequently dead, and moved it for us anyway.  That's why `old_admin.ListTables()` seemed to fix things ... it kept the `BackgroundThreads` alive until the callbacks were complete.

In the end it seems we're best off (as Carlos was probably alluding to) if these objects are created/owned/destroyed at some higher level, and that we could probably favor passing them down by reference rather than by copy to make that even clearer.

**Comment by dbolduc** [comment]
Flake from this issue:

GCB: https://console.cloud.google.com/cloud-build/builds;region=us-east1/ac1135df-2c5e-4605-8a63-f128562c5ffe;tab=detail?project=cloud-cpp-testing-resources
Raw: https://storage.googleapis.com/cloud-cpp-community-publiclogs/logs/google-cloud-cpp/main/508e008ffd7c7a0fbdb092660378464d185f7c8b/fedora-34-coverage/log-ac1135df-2c5e-4605-8a63-f128562c5ffe.txt

**Comment by devbww** [comment]
For the record, in the existing code the `TableAdmin` copy that is destroyed during a callback in a background thread is made in the constructor of `google::cloud::bigtable::TableAdmin::AsyncWaitForConsistencyImpl::AsyncWaitForConsistencyState`, which is allocated by the member `Create()` factory function.  There is no guarantee that this `TableAdmin` copy will not eventually be the one holding the last reference to the `BackgroundThreads`, and therefore it risks trying to join a thread from itself.

**Comment by dbolduc** [comment]
GCB: https://console.cloud.google.com/cloud-build/builds;region=us-east1/59801f7b-077d-4834-893e-df5d720a1596;tab=detail?project=cloud-cpp-testing-resources
Raw: https://storage.googleapis.com/cloud-cpp-community-publiclogs/logs/google-cloud-cpp/main/cd879f427e250318a94cf250bc325da548455540/fedora-34-coverage/log-59801f7b-077d-4834-893e-df5d720a1596.txt
